### PR TITLE
Add engine

### PR DESCRIPTION
**Pull request by mwestphal**
Adding a engine

- public classes are pure abstract, only a header
- implementation of public classes are suffixed _impl
- engine is responsible for creating the correct instances
- Using the static architecture approach is much easier generally, everything it dealt with in constructor, nothing to be done by the user
- Lots and lots of TODO
- A couple of other improvements

### CODE REVIEW

**Comment by mwestphal** [comment]
@Meakk ptal

File: src/library/f3d_options.cxx [hunk 1]
```diff
@@ -155,6 +155,7 @@ options::~options() = default;
 
 //----------------------------------------------------------------------------
 options::options(const options& opt)
+  : Internals(new options::F3DInternals)
```
Comment (mwestphal) [inline src/library/f3d_options.cxx:158]: This is not tested anymore, but it actually should not have been before. I keep the method for now.

File: src/library/f3d_engine.cxx [hunk 2]
```diff
@@ -0,0 +1,230 @@
+#include "f3d_engine.h"
+
+#include "f3d_config.h"
+#include "f3d_init.h" // This includes is needed on mac so that the global variable is created
```
Comment (mwestphal) [inline src/library/f3d_engine.cxx:4]: this was needed for MacOS, not sure about it.

File: src/library/F3DStarter.cxx [hunk 3]
```diff
@@ -64,67 +58,69 @@ int F3DStarter::Start(int argc, char** argv)
...
   }
   else
   {
     // TODO Test this multiconfig behavior
-    this->Internals->Interactor.setKeyPressCallBack(
+    bool offscreen = !this->Internals->CommandLineOptions.Reference.empty() ||
+      !this->Internals->CommandLineOptions.Output.empty();
+
+    this->Internals->Engine =
+      std::make_unique<f3d::engine>(f3d::engine::WindowTypeEnum::WINDOW_STANDARD, offscreen);
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply src/library/F3DStarter.cxx:71]: yeah we need bit flags and im not sure of the right impl, hence the todo

[the remark this answers is not in the document]
Comment (mwestphal) [reply src/library/F3DStarter.cxx:71]: I've put an enum  for now

**Comment by mwestphal** [comment]
@Meakk please review

File: src/library/f3d_window_impl_noRender.cxx [hunk 4]
```diff
@@ -1,27 +1,28 @@
...
 {
 }
 
 //----------------------------------------------------------------------------
-windowNoRender::~windowNoRender() = default;
+window_impl_noRender::~window_impl_noRender() = default;
 
 //----------------------------------------------------------------------------
-vtkRenderWindow* windowNoRender::GetRenderWindow()
+vtkRenderWindow* window_impl_noRender::GetRenderWindow()
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply src/library/f3d_window_impl_noRender.cxx:25]: no vtkRenderWindow, no importers, no importers, no data.

[the remark this answers is not in the document]
Comment (Meakk) [reply src/library/f3d_window_impl_noRender.cxx:25]: Ok I see, this is something to think about, it's awkward to rely on a renWin in this mode.

**Review by Meakk — APPROVED** [review]

File: src/library/f3d_window_impl_noRender.cxx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mwestphal) [reply src/library/f3d_window_impl_noRender.cxx:25]: That something in VTK, not sure we can change that.

**Pull request merged**